File: LeetCode/796-rotate-string/796-rotate-string.go
package main

import (
	"fmt"
	"strings"
)

func rotateString(s string, goal string) bool {
	m := len(s)
	n := len(goal)
	if m > n {
		return false
	}
	if strings.Contains(s+s, goal) {
		return true
	}
	return false
}

// func rotateString(s string, goal string) bool {
// 	m := len(s)
// 	n := len(goal)
// 	if m > n {
// 		return false
// 	}
// 	for i := 0; i < m; i++ {
// 		if s[i:]+s[:i] == goal {
// 			return true
// 		}
// 	}
// 	return false
// }

func main() {
	inputs1 := []string{"abcde", "abcde", "aa"}
	inputs2 := []string{"cdeab", "abced", "a"}
	outputs := []bool{true, false, false}

	for pos := 0; pos < len(outputs); pos++ {
		ans := rotateString(inputs1[pos], inputs2[pos])
		if ans != outputs[pos] {
			fmt.Println(inputs1[pos], inputs2[pos], ans, outputs[pos])
		}
	}

}

File: LeetCode/796-rotate-string/796-rotate-string.rs
struct Solution {}

impl Solution {
    pub fn rotate_string(s: String, goal: String) -> bool {
        let m = s.len();
        let n = goal.len();

        if m > n {
            return false;
        }
        let double_s = format!("{s}{s}");
        return double_s.contains(&goal);
    }
}

fn main() {
    let s10 = &String::from("abcde");
    let s11 = &"abcde".to_string();
    let s12 = &"aa".to_owned();
    let inputs1 = vec![s10, s11, s12];

    let s20 = &"cdeab".to_string();
    let s21 = &"abced".to_string();
    let s22 = &"a".to_string();
    let inputs2 = vec![s20, s21, s22];

    let outputs = vec![true, false, false];

    for pos in 0..outputs.len() {
        let ans = Solution::rotate_string(inputs1[pos].clone(), inputs2[pos].clone());

        if ans != outputs[pos] {
            println!(
                "{:?}, {:?}, {:?}, {:?}",
                inputs1[pos], inputs2[pos], ans, outputs[pos]
            );
        }
    }
}
